tree/ntuple/v7/src/RClusterPool.cxx

```diff
@@ -0,0 +1,325 @@
+            continue;
+         }
+
+         auto cptr = itr->fFuture.get();
+         // If cptr is nullptr, the cluster expired previously and was released by the I/O thread
+         if (!cptr || itr->fIsExpired) {
+            cptr.reset();
+            itr = fInFlightClusters.erase(itr);
+            continue;
+         }
+
+         // We either put a fresh cluster into a free slot or we merge the cluster with an existing one
+         auto existingCluster = FindInPool(cptr->GetId());
+         if (existingCluster) {
+            existingCluster->MergeCluster(std::move(*cptr));
+         } else {
+            auto idxFreeSlot = FindFreeSlot();
+            fPool[idxFreeSlot] = std::move(cptr);
+         }
+         itr = fInFlightClusters.erase(itr);
+      }
+
+      // Deterime clusters which get triggered for background loading
+      for (auto &cptr : fPool) {
+         if (!cptr)
+            continue;
+         provide.Erase(cptr->GetId(), cptr->GetAvailColumns());
+      }
+
+      // Update the work queue and the in-flight cluster list with new requests
+      // TODO(jblomer): we should ensure that clusterId is given first to the I/O thread.  That is usually the
+      // case but it's not ensured by the code
+      std::unique_lock<std::mutex> lockWorkQueue(fLockWorkQueue);
+      for (const auto &kv : provide) {
+         R__ASSERT(!kv.second.empty());
+
+         RWorkItem workItem;
+         workItem.fClusterId = kv.first;
+         workItem.fColumns = kv.second;
+
+         RInFlightCluster inFlightCluster;
+         inFlightCluster.fClusterId = kv.first;
+         inFlightCluster.fColumns = kv.second;
+         inFlightCluster.fFuture = workItem.fPromise.get_future();
+         fInFlightClusters.emplace_back(std::move(inFlightCluster));
+
+         fWorkQueue.emplace(std::move(workItem));
+      }
+      if (fWorkQueue.size() > 0)
+         fCvHasWork.notify_one();
+   } // work queue and in-flight clusters lock guards
+
+   return WaitFor(clusterId, columns);
+}
+
+
+std::shared_ptr<ROOT::Experimental::Detail::RCluster>
+ROOT::Experimental::Detail::RClusterPool::WaitFor(
+   DescriptorId_t clusterId, const RPageSource::ColumnSet_t &columns)
+{
+   while (true) {
+      // Fast exit: the cluster happens to be already present in the cache pool
+      auto result = FindInPool(clusterId);
+      if (result) {
+         bool hasMissingColumn = false;
+         for (auto cid : columns) {
+            if (result->ContainsColumn(cid))
+               continue;
+
+            hasMissingColumn = true;
+            break;
+         }
+         if (!hasMissingColumn)
+            return result;
+      }
+
+      // Otherwise the missing data must have been triggered for loading by now, so block and wait
+      auto itr = fInFlightClusters.begin();
+      for (; itr != fInFlightClusters.end(); ++itr) {
+         if (itr->fClusterId == clusterId)
+            break;
+      }
+      R__ASSERT(itr != fInFlightClusters.end());
```

Comment:
Where is the case "user request completely out-of-band entry" (i.e. random order read request) get caught/stopped/handled?

Reply:
The `GetCluster()` method makes sure that the requested cluster id is at least triggered for loading if it is not anyway already in the cache. So `WaitFor(clusterId)` must find the cluster eventually in the pool and functionally random access does work (I added a [unit test](https://github.com/root-project/root/pull/5377/commits/717d18295b8e905ccbdfb86e55321efa3640a933#diff-2417c332791ee64d3ca77c968cedc592R880)).  It is not very efficient though. It will constantly trigger loading of the following clusters, which then get discarded when the next request for a completely different entry / cluster comes in.

If it is known upfront that the read pattern is random, I think we / the user should just turn off the cluster cache. There is currently no efficient handling of the occasional out-of-band request.  If this is necessary, I think it should be handled in the page source, which can load data directly without the cluster pool.